### DISCUSSION TITLE
[RFC] Component level alias token override

### DISCUSSION THREAD

**Discussion by zombieJ** [ant-design/ant-design#37212] [RFCs - archive]
# 📋 RFC

## Summary

ConfigProvider support component level alias token override. 🚨 Seed, Map, Alias Token should be unique name instead 🚨.

## Basic example

```tsx
<ConfigProvider theme={{
  token: {
    colorPrimary: 'red',  // Seed token override
    colorPrimaryBase: 'blue',  // Alias token override
  },
  components: {
    Dropdown: {
      zIndexPopup: 1234, // Component token override
      colorPrimary: 'green', // Seed token override
    },
  },

  // Other props
  algorithm: darkThemeAlgorithm,
  hashed: false,
}} />
```

## Motivation

It's not easy for developer need only modify one component alias token. They have to HOC component with ConfigProvider wrapper for the override token.

## Detailed

Extends current `useToken` ComponentToken part and add `override` definition. It will auto handle by `@ant-design/cssinjs` lib.

## Drawbacks

Flexible may cause more understand cost.

## Adoption strategy

Dev in `next` branch and release in v5.

**Comment by arvinxx** [comment]
我建议按设计语义来处理？
```tsx
<ConfigProvider
    theme={{
      // global 影响所有组件
      global: {
        colorPrimary: 'green',
      },
      // components 全局变量，只针对组件进行覆盖
      components: {
        Button: {
          colorPrimary: 'blue',
        },
      },
    }}
  />
```

**Reply by zombieJ** [reply]
`override` 全局替换是已经存在的定义，这里只讨论 component 级别的替换：
```tsx
{
  token: { colorPrimary },
  override: { colorPrimaryHover }, // Map Token & Alias Token
  components: {
    Dropdown: {
      zIndexPopup: 123,
      override, // 🔥
    },
  }, 
}
```

这里和 @MadCcc 讨论是否需要打平和 ComponentToken 上一致，目前还是想分的开一些

**Reply by arvinxx** [reply]
override 现在不是还有可以配置Button啥的么，是想把 override 里的抽出来？

**Reply by arvinxx** [reply]
如果有两个 override ，我感觉会让普通开发者迷糊的

**Reply by zombieJ** [reply]
Basic:
```tsx
{
  token: { colorPrimary },
}
```

要进阶才会加 `override`

**Reply by MadCcc** [reply]
```ts
{
  seed: { colorPrimary },
  map: { colorPrimaryBg },
  alias: { controlOutline },
  components: {
    Button: {
      xxxToken: { butonXXX }
      alias: { ... }
    }
  }
}
```

**Reply by zombieJ** [reply]
```diff
  {
    token,

-   override,
+   mapToken,
+   aliasToken,

    components: {
      Dropdown: {
        zIndexPopup: 123123,
        aliasToken: xxxx,
      },
    },
  }
```

Component 级别 seed 和 map 同下，且暂时先不加。

**Reply by dancerphil** [reply]
```diff
  {
+   seedToken,
+   mapToken,
+   aliasToken,

    components: {
      Dropdown: {
        zIndexPopup: 123123,
+       seedToken,
+       mapToken,
+       aliasToken,
      },
    },
  }
```

这样或许可以有些逻辑复用？

**Comment by mushan0x0** [comment]
那这个是不是可以作为覆盖`props`的功能，每个组件都有个`theme`的`prop`

```tsx
<ConfigProvider
  components={{
    Button: {
      style: {
        ...// 用来设置没有token的样式属性
      },
      size: 'large',
      theme: {
        colorPrimary: 'green',
      },
    },
  }}
/>
```

**Reply by dancerphil** [reply]
theme 作为一个整体比较好，这样第三方主题统一提供 theme 就行

**Reply by mushan0x0** [reply]
> theme 作为一个整体比较好，这样第三方主题统一提供 theme 就行

那再提供一种数组的参数类型？
```tsx
<ConfigProvider
  overrides={[darkOverrides, { // 后面的可以覆盖前面的
    Button: {
      style: {
        ...// 用来设置没有token的样式属性
      },
      size: 'large',
      theme: {
        colorPrimary: 'green',
      },
    },
  }]}
/>
```

**Reply by MadCcc** [reply]
这种就没有必要了，能合成一个主题没有必要把两个都传进去

**Reply by mushan0x0** [reply]
> 这种就没有必要了，能合成一个主题没有必要把两个都传进去

这样可以在别的主题上再`mix`感觉会很实用吧，不过这里讨论的是`token`，这样的功能就看后面有没有必要提供吧

**Reply by dancerphil** [reply]
这是另一个需求，在于我想要所有的 Button 默认 size="large"，或者所有 Button 默认 type="primary"。

确实是一个场景，之前没理解到，不过这就不在这个 RFC 的范畴内了。

**Comment by afc163** [comment]
我建议 override 这种作为行为描述就不需要透出在 API 里了，多一层也没啥意义。

**Reply by afc163** [reply]
第一直觉，也是最简单的 API 格式是：

```js
<ConfigProvider
  theme={{
    colorPrimary: '#xxx',  // 全局，也和 antd v4 保持一致
    Button: {
      colorPrimary: '#xxx',  // 组件内的主题覆盖
    },
  }}
/>
```

**Reply by zombieJ** [reply]
这样会有问题，因为 `colorPrimary` 在 seed、map、alias token 都存在。你是没有办法指定覆盖的是哪一层的。比如说，我只想改 alias 里的 `colorPrimary` 而不影响 map 里 `colorPrimary` 派生的 `colorLink`。改了 `colorPrimary` 会导致 `colorLink` 也会变化。

在之前讨论里，我们有聊过 seed、map、alias 里使用不同的名字。比如 seed: colorBrand、map：colorPrimary、alias： colorPrimaryBase。但是讨论下来不希望同一个 token 有太多别名，所以决定在不同 token 里都是用相同的 token 名。那么在现在的覆盖里，就需要明确区分到底这个 token 是哪一层来覆盖的。

**Reply by afc163** [reply]
用户不应该知道和理解 seed、map、alias 这些概念和分层，会搞糊涂的，直觉就是 v4 这种变量。

**Reply by afc163** [reply]
应该就是你 antd 告诉我有哪些变量可以改（应该就是个一维的可枚举的变量列表），改了预期的效果是啥，然后提供一个 API 让我改全局或者改一个组件下的效果就行了。

我不应该去知道有 seed、map、alias 这些东西，也不应该去了解不同层下对应的变量是啥，这样学习成本太高了。对 antd 来说，维护和情况交叉的成本也太高了。

**Reply by mushan0x0** [reply]
`theme`可以考虑做成数组类型参数吗，这样靠先后顺序来覆盖

```tsx
<ConfigProvider
  theme={[darkTheme, { // 跟前面的主题做mix处理
    colorPrimary: '#xxx',  // 全局，也和 antd v4 保持一致
    Button: {
      colorPrimary: '#xxx',  // 组件内的主题覆盖
    },
  }]}
/>
```

**Reply by dancerphil** [reply]
> 比如说，我只想改 alias 里的 `colorPrimary` 而不影响 map 里 `colorPrimary` 派生的 `colorLink`。改了 `colorPrimary` 会导致 `colorLink` 也会变化。

这种情况同时指定两个变量就可以了。

你想的无非是一个算法失效的问题，但 `{seed: {colorPrimary: 'red'}, alias: {colorPrimary: 'pink'}}` 也算是比较精神分裂的了

**Reply by zombieJ** [reply]
> 这种情况同时指定两个变量就可以了。

这个就是之前讨论的，是否不同 Token 需要用不同变量名字。 结论是为了降低记忆成本，不额外添加 token。
